clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/handlers/LoadBalancerUpsertHandler.groovy

```diff
@@ -0,0 +1,115 @@
+import com.amazonaws.services.elasticloadbalancing.model.CreateLoadBalancerRequest
+import com.amazonaws.services.elasticloadbalancing.model.DeleteLoadBalancerListenersRequest
+import com.amazonaws.services.elasticloadbalancing.model.Listener
+import com.amazonaws.services.elasticloadbalancing.model.LoadBalancerDescription
+import com.netflix.spinnaker.clouddriver.data.task.Task
+import com.netflix.spinnaker.clouddriver.data.task.TaskRepository
+import com.netflix.spinnaker.clouddriver.orchestration.AtomicOperationException
+
+class LoadBalancerUpsertHandler {
+
+  private static final String BASE_PHASE = "UPSERT_ELB"
+
+  private static Task getTask() {
+    TaskRepository.threadLocalTask.get()
+  }
+
+  public static void updateLoadBalancer(AmazonElasticLoadBalancing loadBalancing, LoadBalancerDescription loadBalancer,
+                                  List<Listener> listeners, Collection<String> securityGroups) {
+    def amazonErrors = []
+    def loadBalancerName = loadBalancer.loadBalancerName
+
+    if (listeners) {
+      def existingListeners = loadBalancer.listenerDescriptions*.listener.findAll {
+        // ignore the old listener :0 => :0, which AWS adds to ELBs created sometime before 2012-09-26
+        it.protocol && it.instancePort != 0 && it.loadBalancerPort != 0
+      }
+      def listenersToRemove = existingListeners.findAll {
+        // existed previously but were not supplied in upsert and should be deleted
+        !listeners.contains(it)
+      }
+      listeners.removeAll(listenersToRemove)
+
+      // no need to recreate existing listeners
+      listeners.removeAll(existingListeners)
+
+      listenersToRemove.each {
+        loadBalancing.deleteLoadBalancerListeners(
+          new DeleteLoadBalancerListenersRequest(loadBalancerName, [it.loadBalancerPort])
+        )
+        task.updateStatus BASE_PHASE, "Listener removed from ${loadBalancerName} (${it.loadBalancerPort}:${it.protocol}:${it.instancePort})."
+      }
+
+      // ignore the old listener :0 => :0, which AWS adds to ELBs created sometime before 2012-09-26
+      listeners
+        .findAll { it.instancePort != 0 && it.loadBalancerPort != 0 && it.protocol }
+        .each { Listener listener ->
+          try {
+            loadBalancing.createLoadBalancerListeners(new CreateLoadBalancerListenersRequest(loadBalancerName, [listener]))
+            task.updateStatus BASE_PHASE, "Listener added to ${loadBalancerName} (${listener.loadBalancerPort}:${listener.protocol}:${listener.instancePort})."
+          } catch (AmazonServiceException e) {
+            def exceptionMessage = "Failed to add listener to ${loadBalancerName} (${listener.loadBalancerPort}:${listener.protocol}:${listener.instancePort}) - reason: ${e.errorMessage}."
+            task.updateStatus BASE_PHASE, exceptionMessage
+            amazonErrors << exceptionMessage
+          }
+        }
+    }
+
+    if (securityGroups) {
```

Comment:
We should do this first on the off chance that something fails in executing this operation, it would be better to have the security groups in place (thinking of a potential case where a new listener gets added and the corresponding security group update would have some how restricted access to that listener)